File: python/mundane/_duration.py
"""Parse human-friendly duration strings.

Accepts strings like "30s", "5m", "2h", "1d", "500ms", or ints/floats in
seconds (matching neither TS nor JS conventions strictly — the TS API
documents number-of-milliseconds, the Python API documents number-of-seconds).
For consistency with the TS interface, we accept both: strings and numbers
where numbers are treated as milliseconds when called from the public API
(see core.py).

This module only parses strings and returns milliseconds.
"""

import re

_UNIT_MS = {
    "ms": 1,
    "s": 1000,
    "m": 60 * 1000,
    "h": 60 * 60 * 1000,
    "d": 24 * 60 * 60 * 1000,
}

# Matches: "5m", "1h", "2.5s", "500ms". Allow integer or decimal magnitudes.
_RE = re.compile(r"^\s*(\d+(?:\.\d+)?)(ms|s|m|h|d)\s*$", re.IGNORECASE)


def parse_duration_ms(s: str) -> int:
    """Parse a duration string into integer milliseconds."""
    if not isinstance(s, str):
        raise TypeError(f"duration must be a string, got {type(s).__name__}")
    m = _RE.match(s)
    if not m:
        raise ValueError(
            f"invalid duration {s!r}: expected e.g. '500ms', '30s', '5m', '2h', '1d'"
        )
    magnitude = float(m.group(1))
    unit = m.group(2).lower()
    return int(round(magnitude * _UNIT_MS[unit]))

File: go/mundane/duration.go
package mundane

import (
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
)

var durationRe = regexp.MustCompile(`^\s*(\d+(?:\.\d+)?)(ms|s|m|h|d)\s*$`)

var unitMs = map[string]int64{
	"ms": 1,
	"s":  1000,
	"m":  60 * 1000,
	"h":  60 * 60 * 1000,
	"d":  24 * 60 * 60 * 1000,
}

// ParseDurationMs parses "30s", "5m", "2h", "500ms", "1d", "2.5s" into ms.
func ParseDurationMs(s string) (int64, error) {
	m := durationRe.FindStringSubmatch(strings.ToLower(s))
	if m == nil {
		return 0, fmt.Errorf("invalid duration %q: expected e.g. '500ms', '30s', '5m', '2h', '1d'", s)
	}
	mag, err := strconv.ParseFloat(m[1], 64)
	if err != nil {
		return 0, fmt.Errorf("invalid duration magnitude %q: %w", m[1], err)
	}
	unit, ok := unitMs[m[2]]
	if !ok {
		return 0, fmt.Errorf("invalid duration unit %q", m[2])
	}
	return int64(math.Round(mag * float64(unit))), nil
}
